homeassistant/components/ecovacs/config_flow.py

```diff
@@ -21,23 +25,39 @@
 _LOGGER = logging.getLogger(__name__)
 
 
-def validate_input(user_input: dict[str, Any]) -> dict[str, str]:
+async def _validate_input(
+    hass: HomeAssistant, user_input: dict[str, Any]
+) -> dict[str, str]:
     """Validate user input."""
     errors: dict[str, str] = {}
+
+    deebot_config = Configuration(
+        aiohttp_client.async_get_clientsession(hass),
+        device_id=get_client_device_id(),
+        country=user_input[CONF_COUNTRY],
+        continent=user_input.get(CONF_CONTINENT),
+    )
+
+    authenticator = Authenticator(
+        deebot_config,
+        user_input[CONF_USERNAME],
+        md5(user_input[CONF_PASSWORD]),
+    )
+
     try:
-        EcoVacsAPI(
-            get_client_device_id(),
-            user_input[CONF_USERNAME],
-            EcoVacsAPI.md5(user_input[CONF_PASSWORD]),
-            user_input[CONF_COUNTRY],
-            user_input[CONF_CONTINENT],
-        )
-    except ValueError:
+        await authenticator.authenticate()
+    except ClientError:
+        _LOGGER.debug("Cannot connect", exc_info=True)
+        errors["base"] = "cannot_connect"
+    except InvalidAuthenticationError:
         errors["base"] = "invalid_auth"
     except Exception:  # pylint: disable=broad-except
-        _LOGGER.exception("Unexpected exception")
+        _LOGGER.exception("Unexpected exception during login")
         errors["base"] = "unknown"
 
+    if (continent := user_input.get(CONF_CONTINENT)) and len(continent) != 2:
+        errors["CONF_CONTINENT"] = "invalid_continent_length"
```

Comment:
Can't you use config validators for this? (I mean I saw you are implementing a nicer error message for those)

Reply:
The error messages from `voluptuous` are not translatable and only available in English.
@MartinHjelmare What's your opinion on it? Should I use the `vol.Length` validator or leave the custom check above?

Reply:
Doesn't https://github.com/home-assistant/core/pull/108075 provide something for this?

Reply:
Removed continent completely